sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/aad/webapi/AADOAuth2OboAuthorizedClientRepository.java

```diff
@@ -0,0 +1,135 @@
+import com.nimbusds.jwt.JWT;
+import com.nimbusds.jwt.JWTParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.oauth2.client.OAuth2AuthorizedClient;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.web.OAuth2AuthorizedClientRepository;
+import org.springframework.security.oauth2.core.AbstractOAuth2Token;
+import org.springframework.security.oauth2.core.OAuth2AccessToken;
+import org.springframework.security.oauth2.server.resource.authentication.AbstractOAuth2TokenAuthenticationToken;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.net.MalformedURLException;
+import java.time.Instant;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+public class AADOAuth2OboAuthorizedClientRepository implements OAuth2AuthorizedClientRepository {
+
+    private static final Logger LOG = LoggerFactory.getLogger(AADOAuth2OboAuthorizedClientRepository.class);
+
+    private final AzureClientRegistrationRepository azureClientRegistrationRepository;
+
+    private Map<String, ConfidentialClientApplication> confidentialClientApplicationMap = new HashMap<>();
+
+    public AADOAuth2OboAuthorizedClientRepository(AzureClientRegistrationRepository azureClientRegistrationRepository) {
+        this.azureClientRegistrationRepository = azureClientRegistrationRepository;
+        Iterator<ClientRegistration> iterator = azureClientRegistrationRepository.iterator();
+        while (iterator.hasNext()) {
+            ClientRegistration next = iterator.next();
+            this.confidentialClientApplicationMap.put(next.getClientId(), createApp(next));
+        }
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T extends OAuth2AuthorizedClient> T loadAuthorizedClient(String registrationId,
+                                                                     Authentication authentication,
+                                                                     HttpServletRequest request) {
+        try {
+            ClientRegistration clientRegistration = azureClientRegistrationRepository
+                .findByRegistrationId(registrationId);
+
+            AbstractOAuth2TokenAuthenticationToken<AbstractOAuth2Token> authenticationToken =
+                (AbstractOAuth2TokenAuthenticationToken)
+                    authentication;
+
+            String accessToken = authenticationToken.getToken().getTokenValue();
+
+            OnBehalfOfParameters parameters = OnBehalfOfParameters
+                .builder(clientRegistration.getScopes(), new UserAssertion(accessToken))
+                .build();
+
+            ConfidentialClientApplication clientApplication = confidentialClientApplicationMap.get(clientRegistration
```

Comment:
I'm not familiar with Spring Security - are you creating a ConfidentialClient application per request, or is one created at startup and then reused?

Reply:
It's created at startup and cached in the map for reusing.

Reply:
Keep in mind that even though you are not using the token cache, MSAL stores tokens into an in-memory cache by default. If the `ConfidentialClientApplication` is reused across thousands of users (or more), the tokens for these users will be always be in the cache. 

If you then actually plan on using the token cache, this will be a problem, as the look up logic for tokens runs in linear time. Therefore, it is recommended that you create a `ConfidentialClientApplication` (and therefore cache) per user.

Reply:
Thanks for the reminder, is there a way to disable the auto-creation of the MSAL token cache? As we're not going to use MSAL token cache, the action of caching the `ConfidentialClientApplication` could be a memory problem. Will the token stored in memory expire ever?

Reply:
Unfortunately there is no way to disable auto-creation of the MSAL token cache. It is highly recommended that customers use cached tokens for many reasons, including end-user experience, developer experience, as well as to reduce load on AAD service. 

Tokens stored in cache due expire, but are not evicted (expired tokens will remain in the cache, unless you remove them, by calling `RemoveAccounts`). If you use the acquireTokenSilent API, MSAL automatically takes care of refreshing the expired tokens for you.

Reply:
@sangonzal thanks for the detailed explanation. Since in our web API case, the application would mostly be sessionless, so we're not using the session to cache token. With your explanation, I'm thinking of three ways to handle the MSAL token cache:
||Pros|Cons|
|--|--|--|
| Create `ConfidentialClientApplication` for each request and call `acquireToken()` | No need to worry about the memory problem.  |- The cost of construct the object. <br/> - The cost of acquiring a token each time.|
| Cache the `ConfidentialClientApplication` at the application level and call `acquireToken()`| Save the cost of constructing `ConfidentialClientApplication`  at each request.  | - All tokens acquired within the application lifecycle will be cached in the MSAL token cache |
| Cache the `ConfidentialClientApplication` at the application level and call `acquireTokenSilent`| - Save the cost to acquire token each time. <br/>- Automatically refresh tokens. <br/> - Save the cost of constructing `ConfidentialClientApplication`  at each request.  | - All tokens will be cached in memory <br/> - Code is not self-explanatory (need to perform obo first, check [this](https://github.com/Azure-Samples/ms-identity-java-webapi/blob/master/msal-obo-sample/src/main/java/com/microsoft/azure/msalobosample/MsalAuthHelper.java#L61-L76))|

But one of my concerns is considering the distributing nature of Web API, what's the benefit of caching the token in the application? Which solution do you prefer among these three, or do you have other suggestion?